### PR TITLE
Offload background integrity_check off the main event loop

### PR DESCRIPTION
**Pull request by gsxdsm**
## What

Follow-up to #1692 (already merged). That PR bounded the WAL and fixed the `vacuum()` lock release; this one removes the **largest remaining event-loop stall** found in the same read-contention investigation.

`scheduleBackgroundIntegrityCheck` (~60s after init) ran `PRAGMA integrity_check` on the live connection, walking every page and **freezing the event loop for several seconds**.

- Offload it to the `sqlite3` CLI in a child process (async `spawn`), matching the existing out-of-process pattern (`quickCheckSqliteFile`, `.recover`).
- Open the CLI connection **`-readonly`** so it can never checkpoint or write the live WAL — verified this works on a live DB (the process holds the connection open, so `-shm` exists) and fails cleanly otherwise.
- Fall back to the in-process check when the CLI is unavailable / can't open read-only (`verified=false`) — preserving today's behavior on those environments.
- New `runBackgroundIntegrityCheck()` seam centralizes the offload+fallback policy and gives the scheduler one deterministic, testable point; the scheduler callback is now a guarded async IIFE so a background timer can't crash the process.

**VACUUM is intentionally not offloaded:** the call graph shows it runs only via the `fn db vacuum` CLI command and tests — never the periodic maintenance loop — so it is not a background stall, and an out-of-process VACUUM on a live WAL DB would add corruption surface for no hot-path benefit.

## Tests
- New coverage for `integrityCheckSqliteFileAsync` (healthy live DB + non-existent file).
- The 4 startup-integrity tests updated to the async/offloaded seam, kept deterministic regardless of whether the `sqlite3` CLI exists in CI.
- Affected suites pass; `tsc --noEmit` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://stagereview.app/Runfusion/Fusion/pull/1693">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Upgraded database integrity verification for disk-backed databases using an external sqlite integrity check, returning richer status details (including verification outcome and error reasons).

* **Bug Fixes**
  * Improved background integrity check reliability: failures or rejected runs now consistently clear pending state across participating databases, including safeguards for closed instances.
  * Added a safer fallback path when the external checker can’t be used.

* **Tests**
  * Expanded integrity-check test coverage, including healthy databases, missing files, rejection/throw edge cases, and correct async background completion.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 4 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Implement async SQLite integrity check utility](https://stagereview.app/Runfusion/Fusion/pull/1693/chapters/1) |
| 2 | [Add background integrity check orchestration logic](https://stagereview.app/Runfusion/Fusion/pull/1693/chapters/2) |
| 3 | [Wire async integrity check into scheduler](https://stagereview.app/Runfusion/Fusion/pull/1693/chapters/3) |
| 4 | [Update and expand database integrity tests](https://stagereview.app/Runfusion/Fusion/pull/1693/chapters/4) |

<a href="https://stagereview.app/Runfusion/Fusion/pull/1693">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit 580e72b6a7f2ae35a241463c3c510f523c1748ad on Jun 20, 2026 9:35pm UTC.</sub>

File: packages/core/src/db.ts [hunk 1]
```diff
@@ -1993,6 +2075,27 @@ export class Database {
...
+  private async runBackgroundIntegrityCheck(): Promise<{ ok: true } | { ok: false; errors: string[] }> {
+    if (this.inMemory) {
+      return this.integrityCheck();
+    }
+    const offloaded = await integrityCheckSqliteFileAsync(this.dbPath);
+    if (offloaded.verified) {
+      return offloaded.ok ? { ok: true } : { ok: false, errors: offloaded.errors ?? [] };
+    }
+    return this.integrityCheck();
+  }
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/db.ts:2128]: Fixed in 580e72b6a. Confirmed real and reproduced by independent review (`prepare()` does throw on a closed `DatabaseSync`).

Two-part fix:
1. `runBackgroundIntegrityCheck` now guards `this.closed` before **every** in-process `integrityCheck()` call — both before the offload and again after the `await` (the connection can close during the seconds-long CLI run).
2. The scheduler now clears `integrityCheckPending` **unconditionally in a `finally`**, so even if the check throws, no participant is left pinned at `pending=true`. Added a regression test asserting pending clears for all participants when the check throws.

File: packages/core/src/db.ts [hunk 2]
```diff
@@ -5277,30 +5380,43 @@ export class Database {
...
+            `[fusion:db] Background integrity check detected corruption for ${this.dbPath}: ${errorSummary}`,
+          );
+        }
+      })()
+        .catch((error) => {
+          console.warn(`[fusion:db] Background integrity check failed for ${this.dbPath}`, error);
+        })
+        .finally(() => {
+          Database.sharedIntegrityChecks.delete(this.dbPath);
+        });
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/db.ts:5461]: Fixed in 580e72b6a. The clearing loop now re-reads the **current** `shared.subscribers` set inside the `finally` (after the await) instead of a pre-await snapshot, so a `Database` that subscribed during the running window is cleared too. Since the loop runs synchronously immediately before the `.finally()` `delete`, the late-subscriber window is back to ~zero (no `await` between clear and delete).

File: packages/core/src/db.ts [hunk 3]
```diff
@@ -5277,30 +5380,43 @@ export class Database {
...
-        participant.corruptionDetected = !integrity.ok;
-        participant.integrityCheckErrors = integrity.ok ? [] : [...integrity.errors];
-      }
+      // FNXC:Database 2026-06-20-13:30:
+      // Offload the integrity-check page-walk to the sqlite3 CLI in a child
+      // process so it no longer blocks the event loop for several seconds. The
+      // in-process check (primary.integrityCheck()) remains the fallback for
+      // environments without the sqlite3 CLI. Wrapped in an async IIFE because
+      // setTimeout callbacks can't be async; errors must be swallowed here so an
+      // unhandled rejection can't crash the process from a background timer.
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/db.ts:5420]: Wont fix — keeping the marker. `FNXC:` is a **mandatory project convention** documented in `AGENTS.md` (FNXC_LOG comments): agent-authored comments encoding a requirement must carry the `FNXC:Area yyyy-MM-dd-hh:mm:` prefix. A prior review on this codebase actually flagged the *absence* of these markers as a defect, so removing them would regress the standard. (I did add the missing markers to the two new symbols in this commit for consistency.)
